src/cls/rbd/cls_rbd.cc

```diff
@@ -8699,6 +8699,81 @@ int rwlcache_free(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
   return 0;
 }
 
+/*
+ * ping from librbd primary
+ *
+ * Input
+ * @param cache id (epoch_t)
+ *
+ * Output
+ * @param has removed daemons (bool)
+ * @return 0 on success, negative error code on failure
+ */
+int rwlcache_primaryping(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  epoch_t cache_id;
+  cls_rbd_rwlcache_map map;
+  int r;
+
+  try {
+    auto it = in->cbegin();
+    decode(cache_id, it);
+  } catch (const ceph::buffer::error &err) {
+    return -EINVAL;
+  }
+  CLS_LOG(20, "primaryping with cache %u", cache_id);
+
+  bufferlist value;
+  r = cls_cxx_map_get_val(hctx, RWLCACHE_MAP_KEY, &value);
+  if (r < 0) {
+    CLS_ERR("error reading key %s: %s", RWLCACHE_MAP_KEY, cpp_strerror(r).c_str());
+    return r;
+  }
+
+  try {
+    auto it = value.cbegin();
+    decode(map, it);
+  } catch (const ceph::buffer::error &err) {
+    CLS_ERR("decode map error");
+    return -EIO;
+  }
+
+  bool has_removed_daemon = false;
+  auto cache = map.caches.find(cache_id);
+  if (cache != map.caches.end()) {
+    for (auto id : cache->second.daemons) {
+      if (map.daemons.count(id) == 0) {
+        has_removed_daemon = true;
+        break;
+      }
+    }
+  } else {
+    auto uncommitted_cache = map.uncommitted_caches.find(cache_id);
```

Comment:
what would cause this phenomenon? Can you give a example? From my point, only after send primary_ack, we start sending primary_ping.

Also, if daemon found ping timeout, it should do error handle, etc close all active rdma_connections. Primary handle dis-connection event. When do we need this command?